contracts/SlashIndicator.sol

```diff
@@ -158,30 +180,115 @@ contract SlashIndicator is ISlashIndicator,System,IParamSubscriber, IApplication
     emit indicatorCleaned();
   }
 
+  function submitFinalityViolationEvidence(FinalityEvidence calldata _evidence) external onlyInit onlyRelayer {
+    if (finalityDistance == 0 || finalitySlashRewardRatio == 0) {
+      finalityDistance = FINALITY_DISTANCE;
+      finalitySlashRewardRatio = FINALITY_SLASH_REWARD_RATIO;
+    }
+
+    require(
+      (_evidence.numA < _evidence.numB && _evidence.numA + finalityDistance >= _evidence.numB) ||
+      (_evidence.numB < _evidence.numA && _evidence.numB + finalityDistance >= _evidence.numA),
+      'too long distance between blocks'
+    );
+    require(
+      _evidence.numA < block.number && _evidence.numA + 256 > block.number &&
+      _evidence.numB < block.number && _evidence.numB + 256 > block.number,
+      'block number out of range'
+    );
+
+    // BLS verification
+    (address[] memory vals, bytes[] memory voteAddrs) = IBSCValidatorSet(VALIDATOR_CONTRACT_ADDR).getLivingValidators();
+    bytes memory voteAddress;
+    bool exist;
+    for (uint i = 0; i < vals.length; i++) {
+      if (vals[i] == _evidence.valAddr) {
+        exist = true;
+        voteAddress = voteAddrs[i];
+        break;
+      }
+    }
+    require(exist, "validator not exist");
+
+    bytes memory input;
+    bytes memory output;
+
+    // to avoid too deep stack
+    {
+      bytes memory bytesNumA;
+      bytes memory bytesNumB;
+      bytes memory bytesHeaderA;
+      bytes memory bytesHeaderB;
+      TypesToBytes.uintToBytes(32, _evidence.numA, bytesNumA);
+      TypesToBytes.bytes32ToBytes(32, _evidence.headerA, bytesHeaderA);
+      TypesToBytes.uintToBytes(32, _evidence.numB, bytesNumB);
+      TypesToBytes.bytes32ToBytes(32, _evidence.headerB, bytesHeaderB);
+
+      input = BytesLib.concat(bytesNumA, bytesHeaderA);
+      input = BytesLib.concat(input, _evidence.sigA);
+      input = BytesLib.concat(input, bytesNumB);
+      input = BytesLib.concat(input, bytesHeaderB);
+      input = BytesLib.concat(input, _evidence.sigB);
+      input = BytesLib.concat(input, voteAddress);
+    }
+
+    assembly {
+      let len := mload(input)
+      if iszero(call(not(0), 0x64, 0, input, len, output, 0x20)) {// precompiled contract address 0x64
+        revert(0, 0)
+      }
+    }
+
+    bytes32 _headerA = blockhash(_evidence.numA);
+    bytes32 _headerB = blockhash(_evidence.numB);
+
+    if (_headerA != _evidence.headerA && _headerB != _evidence.headerB) {
```

Comment:
Let us do the check before calling the precompile contract

Reply:
Fixed